streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java

```diff
@@ -1229,7 +1229,7 @@ private <VR, KO, VO> KTable<K, VR> doJoinOnForeignKey(final KTable<KO, VO> forei
 
         final StreamPartitioner<K, SubscriptionResponseWrapper<VO>> foreignResponseSinkPartitioner =
                 tableJoinedInternal.partitioner() == null
-                        ? (topic, key, subscriptionResponseWrapper, numPartitions) -> subscriptionResponseWrapper.getPrimaryPartition()
+                        ? (topic, key, subscriptionResponseWrapper, numPartitions) -> Optional.of(Collections.singleton(subscriptionResponseWrapper.getPrimaryPartition()))
```

Comment:
Should we change `getPrimaryPartition` to return an `Optional` instead ?

Reply:
Sounds good, should we also create an additional constructor dropping the parameter `primaryPartition`? I see many calls of the actual constructor passing `null`? Something like this?

```java
// New Additional Constructor
public SubscriptionResponseWrapper(final long[] originalValueHash, final FV foreignValue) {
        this(originalValueHash, foreignValue, CURRENT_VERSION, null);
}

// Other Constructors
public SubscriptionResponseWrapper(final long[] originalValueHash, final FV foreignValue, final Integer primaryPartition) {
        this(originalValueHash, foreignValue, CURRENT_VERSION, primaryPartition);
}

public SubscriptionResponseWrapper(
        final long[] originalValueHash,
        final FV foreignValue,
        final byte version,
        final Integer primaryPartition) {
        // ... code omitted 
        this.primaryPartition = Optional.ofNullable(primaryPartition);
}
```

Reply:
I don't think it's worth to add an overloaded constructor. This adds more noise than it simplifies code IMHO.

However, looking into the code mode closely, it seems that `getPrimaryPartition ` is called elsewhere, so changing the return type to `Optional` might actually not be ideal?

However, we need to take into account that `getPrimaryPartition` could return `null` for which we should not return a singleton collection with `null` in it here, but rather an `Optional.empty()` (please correct me if I am wrong).

Reply:
@mjsax, about the overloaded, let put it aside, as you mentioned better keep it simple.

About `getPrimaryPartition` we should use `Optional.empty()` in this case, and `Optional.ofNullable()` for the constructor parameter.

However about the returning type change in `getPrimaryPartition`, I noticed that `SubscriptionResponseWrapper` is serialized, do we should concern about the type change in this case?

Reply:
Yes, let's keep the return type of `getPrimaryPartition` unchanged.

However, if `getPrimaryPartition` returns `null`, we should not construct `Optional.of(Collections.singleton(null)` here, but rather `Optional.empty()`, right? So we need to change the code here accordingly?

Reply:
As a matter of fact, we have a failing test that I believe exactly exposes this bug:
```
java.lang.NullPointerException
	at org.apache.kafka.streams.processor.internals.RecordCollectorImpl.send(RecordCollectorImpl.java:163)
	at org.apache.kafka.streams.processor.internals.RecordCollectorTest.shouldUseDefaultPartitionerAsPartitionReturnsNull(RecordCollectorTest.java:606)
	```

Reply:
Just pushed the change :)